feat: Integrate enabled network map (functionality)

gambinish opened the pull request:
## **Description**

This PR integrates the enabledNetworks state from the NetworkOrderController throughout the application, effectively replacing the legacy tokenNetworkFilter. This transition impacts:

- Asset polling logic. We should poll on chains enabled with `enabledNetworkMap` rather than `tokenNetworkFilter`

- dApp network switching. It should enable networks just as it previously updated `tokenNetworkFilter`.

- Network addition flows. When networks are added, they should be enabled, both manually through existing network picker and via dApp.

The goal is to centralize network enablement state in one place `enabledNetworkMap` and streamline how the app filters tokens and interacts with enabled networks. This change is foundational to supporting the new global network selector.

Next step will be to replace the NetworkFilter entirely with a new component to handle network enablement in a multi-select way.

[Designs for reference](https://www.figma.com/design/CvBEDNwQqfW4Wbil40Bhr1/Substitute-Global-Network-selector?node-id=669-58227&t=nE8tOXpi5kIHuZJu-0)

Note that this PR affected a ton of e2e tests, since many tests are reliant on aggregated balance being loaded, along with certain networks to be enabled to match the currently selected global network for certain assets to show up/filter as expected. I've added this tests separately here: https://github.com/MetaMask/metamask-extension/pull/33407

☝️ I'm open to breaking this up further to differentiate between unit test and e2e test updates. I'm also happy to discuss ways to make this follow up PR more minimal. 

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/33478?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Please test with REMOVE_GNS='true' as well as without the feature flag for regressions
2. While REMOVE_GNS='true'
- NetworkFilter will update enabledNetworks to only poll on networks that have been designated from NetworkFilter. This will eventually become the new network picker/enabler
- dApps can switch networks on behalf of MetaMask. This will enable the network that was switched to in dApp
3 While REMOVE_GNS=''
- All existing behavior of NetworkFilter should remain (meaning that it relies on tokenNetworkFilter), however, enabledNetworkMap will also be updated
- enabledNetworkMap should match tokenNetworkFilter and should follow same logic paths:
1. via manual token filtering on NetworkFilter
2. via manual network switching

## **Screenshots/Recordings**

With `REMOVE_GNS='true` feature flag on:

Network Filter on MM:
https://github.com/user-attachments/assets/ce331108-c588-44bc-9677-8c50128513f1

dApp network switch:
https://github.com/user-attachments/assets/b617b527-029a-4b2f-992f-611ebe11429f

With `REMOVE_GNS=''`

https://github.com/user-attachments/assets/49aee746-6930-41e4-9f70-5a0f905d6a78

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

gambinish committed:
chore: Update ethereum-chain-utils.js

gambinish committed:
chore: Update switch-ethereum-chain.js

gambinish committed:
chore: Update metamask-controller.js

gambinish committed:
chore: Update asset-list-control-bar.tsx

gambinish committed:
chore: Update network-filter.tsx

gambinish committed:
chore: Update defi-list.tsx

gambinish committed:
chore: Update token-list.tsx

gambinish committed:
chore: Update import-tokens-modal.js

gambinish committed:
chore: Update network-list-menu

gambinish committed:
chore: Update asset polling hooks with enabledNetworkMap

gambinish committed:
chore: Update selectors.js

gambinish committed:
chore: Merge main, address conflicts

gambinish committed:
chore: Cleanup

gambinish committed:
chore: Semantically leverage REMOVE_GNS ff

gambinish committed:
chore: Updated ff changes to defi and token lists

gambinish committed:
chore: Update getIsTokenNetworkFilterEqualCurrentNetwork with correct feature flag

gambinish committed:
Merge branch 'main' into feat/integrate-enabledNetworkMap

gambinish commented on ui/components/multichain/network-list-menu/network-list-menu.tsx:
```diff
@@ -362,6 +365,12 @@ export const NetworkListMenu = ({ onClose }: NetworkListMenuProps) => {
     }
 
+    if (Object.keys(enabledNetworks).length === 1) {
+      dispatch(setEnabledNetworks([hexChainId] as CaipChainId[]));
```
Yeah, good callout. It's a hex chainId. The `AssetPollingProvider` expects the chainIds in their polling loops, so they'd have to be converted to a chainId before they hit our assets controllers, else we would need to do a pretty big refactor.

I suppose that they could be stored in the `enabledNetworks` mapping as CaipChainIds, and then converted to hex chainIds in the polling loops. I did it this way instead to make the fewest changes possible, especially considering e2e. But you could argue that this typecasting is an antipattern.

If you think it should be done this way lmk, and I can look to see how big of a change this would be.

micaelae replied:
Does anything consume them as a CaipChainId or do all the consumers expect Hex? I think it would be safer for the types to match to avoid unexpected behavior when using the state values

micaelae replied:
In the bridge experience, we handled this by accepting [all versions/types of the chainId](https://github.com/MetaMask/metamask-extension/blob/main/ui/ducks/bridge/types.ts#L15) (number | Hex | CaipChainId). Conversions are not performed unless a consumer needs a specific type. This way the type-checking can still catch issues and the conversions happen minimally

gambinish replied:
The EVM polling providers all expect hex chainIds, but Solana expects the CaipChainId. Let me think about this a bit more to see if something makes sense here. I'll reach out to you tomorrow on Slack with a more thought out approach.

gambinish committed:
chore: Loosen Caip19+Hex type to string

gambinish committed:
fix: Lint

vinnyhoward commented on ui/store/actions.ts:
```diff
@@ -4459,7 +4459,7 @@ export function updateAccountsList(
  * to be enabled. All other networks will be implicitly disabled.
  */
 export function setEnabledNetworks(
-  chainIds: CaipChainId[],
```
Same question as [this](https://github.com/MetaMask/metamask-extension/pull/33478/files#r2138705702)

vinnyhoward reviewed (dismissed):
LGTM

gambinish committed:
feat: Integrate enabledNetworkMap (tests) (#33407)

## **Description**

This PR updates the unit and end-to-end test coverage for the
functionality introduced in
https://github.com/MetaMask/metamask-extension/pull/33478, which
involved replacing `tokenNetworkFilter` with the new `enabledNetworks`
implementation. While the core functionality touched only ~15 files,
it’s foundational to many areas of the app, requiring many test updates
across 50+ files.

Because of this, I opted to create two separate PRs for ease of review.
One with functionality changes (33478), the other with test updates.
There is also a third which enables the feature with a feature flag,
which we can use to generate QA builds.

The changes in this PR, along with 33478 are required for us to ship the
upcoming Network Enablement UI with the REMOVE_GNS feature flag enabled.
This PR: https://github.com/MetaMask/metamask-extension/pull/33488
proves that it is green on CI with this flag enabled, so we will only
need to create new e2e tests to cover this new functionality.

Key changes:
• Introduced a new `enabledNetworks` test mock to replace legacy
`tokenNetworkFilter` usage.
	•	Updated unit tests to align with the new network filtering logic.
• Updated E2E specs to reflect expected behavior under the new
enablement model.

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/33407?quickstart=1)

## **Related issues**

Functionality PR:
https://github.com/MetaMask/metamask-extension/pull/33478
Feature Flag PR:
https://github.com/MetaMask/metamask-extension/pull/33488

Please let me know via DM if any of this is unclear. Happy to clarify
over a quick call.

## **Manual testing steps**

## **Screenshots/Recordings**

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

gambinish committed:
Merge branch 'main' into feat/integrate-enabledNetworkMap

gambinish committed:
chore: Update mock

gambinish committed:
Merge branch 'main' into feat/integrate-enabledNetworkMap

OGPoyraz reviewed (dismissed):
Confirmation changes LGTM

gambinish committed:
chore: Remove megaeth and monad from privacy-snapshot

gambinish committed:
Merge branch 'feat/integrate-enabledNetworkMap' of github.com:MetaMask/metamask-extension into feat/integrate-enabledNetworkMap
